src/App/Version/TH.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module App.Version.TH
+  ( getCurrentTag,
+  )
+where
+
+import Control.Carrier.Diagnostics (resultValue, runDiagnostics)
+import Control.Effect.Diagnostics
+  ( Diagnostics,
+    fromEitherShow,
+  )
+import qualified Data.ByteString.Lazy as BSL
+import Data.Maybe (fromMaybe)
+import Data.Text (Text)
+import qualified Data.Text.Encoding as TE
+import qualified Data.Text as T
+import Data.Versions
+import Effect.Exec
+import GitHash
+import Instances.TH.Lift ()
+import Language.Haskell.TH
+import Path
+import Language.Haskell.TH.Syntax (addDependentFile)
+import Path.IO (getCurrentDir)
+
+
+gitTagPointCommand :: Text -> Command
+gitTagPointCommand commit =
+  Command
+    { cmdName = "git",
+      cmdArgs = ["tag", "--points-at", commit],
+      cmdAllowErr = Always
+    }
+
+gitLogFile :: Path Rel File
+gitLogFile = $(mkRelFile ".git/logs/HEAD")
+
+getCurrentTag :: TExpQ (Maybe Text)
+getCurrentTag = do
+  let commitHash = giHash $$(tGitInfoCwd)
+  result <- runIO . runDiagnostics . runExecIO . getTags $ T.pack commitHash
+
+  {- addDependentFile will cause this function to recompile when .git/logs/HEAD
+      changes. This is required for accuracy during dev since this value is not
+      always needed in the final binary.
+
+      NOTE that there is poentially a bug with this function, which causes it to
+      do nothing.  https://github.com/haskell/cabal/issues/4746
+  -}
+  logFile <- runIO $ (</> gitLogFile) <$> getCurrentDir
+  addDependentFile $ toFilePath logFile
+
+  case result of
+    Left err -> reportWarning (show err) >> [||Nothing||]
+    Right tags -> filterTags $ resultValue tags
+
+getTags :: (Has Exec sig m, Has Diagnostics sig m) => Text -> m [Text]
+getTags hash = do
+  -- FIXME: boy it would be nice to not shell out during compilation.  So much that could go wrong.
```

Comment:
It's totally viable, but not the right solution at the moment, in my opinion.  There are some difficulties with doing it that way, and this instability is both rare and not affecting customers.  It would only ever affect devs of this project.  We can improve this later once we implement full metadata.